pkg/store/gcs.go

```diff
@@ -0,0 +1,324 @@
+			err = e
+		}
+	}
+	return err
+}
+
+// SyncBlocks synchronizes the stores state with the GCS bucket.
+func (s *GCSStore) SyncBlocks(ctx context.Context) {
+	// NOTE(fabxc): watches are not yet supported by the Go client library so we just
+	// do a periodic refresh.
+	err := runutil.Repeat(60*time.Second, ctx.Done(), func() error {
+		if err := s.syncBlocks(ctx); err != nil {
+			level.Warn(s.logger).Log("msg", "syncing blocks failed", "err", err)
+		}
+		return nil
+	})
+	if err != nil {
+		level.Error(s.logger).Log("msg", "unexpected error", "err", err)
+	}
+}
+
+func (s *GCSStore) syncBlocks(ctx context.Context) error {
+	objs := s.bucket.Objects(ctx, &storage.Query{Delimiter: "/"})
+
+	for {
+		oi, err := objs.Next()
+		if err == iterator.Done {
+			break
+		} else if err != nil {
+			return err
+		}
+		id, err := ulid.Parse(oi.Prefix[:len(oi.Prefix)-1])
+		if err != nil {
+			continue
+		}
+		if _, ok := s.blocks[id]; ok {
+			continue
+		}
+		level.Info(s.logger).Log("msg", "sync block from GCS", "id", id)
+
+		if err := s.syncBlock(ctx, id); err != nil {
+			level.Error(s.logger).Log("msg", "syncing block failed", "err", err, "block", id.String())
+		}
+	}
+	return nil
+}
+
+// loadFromDisk loads a block with the given ID from the disk cache.
+func (s *GCSStore) loadFromDisk(id ulid.ULID) (*gcsBlock, error) {
+	dir := filepath.Join(s.dir, id.String())
+
+	indexr, err := tsdb.NewFileIndexReader(filepath.Join(dir, "index"))
+	if err != nil {
+		return nil, errors.Wrap(err, "open index reader")
+	}
+	b := &gcsBlock{indexr: indexr, dir: dir}
+
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	s.blocks[id] = b
+	return b, nil
+}
+
+// syncBlock downloads the index and meta.json file for the given block ID and opens a reader
+// against the block.
+func (s *GCSStore) syncBlock(ctx context.Context, id ulid.ULID) error {
+	bdir := filepath.Join(s.dir, id.String())
+	tmpdir := bdir + ".tmp"
+
+	if err := os.MkdirAll(tmpdir, 0777); err != nil {
+		return errors.Wrap(err, "create temp dir")
+	}
+
+	indexObj := s.bucket.Object(path.Join(id.String(), "index"))
+
+	f, err := os.Create(filepath.Join(tmpdir, "index"))
+	if err != nil {
+		return errors.Wrap(err, "create local index copy")
+	}
+	r, err := indexObj.NewReader(ctx)
+	if err != nil {
+		return errors.Wrap(err, "create index object reader")
+	}
+	_, err = io.Copy(f, r)
+
+	if err := f.Close(); err != nil {
+		level.Warn(s.logger).Log("msg", "close file", "err", err)
+	}
+	if err := r.Close(); err != nil {
+		level.Warn(s.logger).Log("msg", "close object reader", "err", err)
+	}
+	if err != nil {
+		if err := os.RemoveAll(tmpdir); err != nil {
+			level.Warn(s.logger).Log("msg", "cleanup temp dir after failure", "err", err)
+		}
+		return errors.Wrap(err, "copy index file to disk")
+	}
+
+	if err := renameFile(tmpdir, bdir); err != nil {
+		return errors.Wrap(err, "rename block directory")
+	}
+
+	_, err = s.loadFromDisk(id)
+	return errors.Wrap(err, "load synced block")
```

Comment:
shouldn't we wrap only on error?

Reply:
`errors.Wrap` returns `nil` again if the input error was `nil`. But I can make an explicit `if` if it seems to magical.

Reply:
Hm.. I would vote for explicit check (: